Allow `retries` without `until` to retry until task is successful

lilatomic opened the pull request:
##### SUMMARY

Specifying `retries` without `until` defaults to using the `{{ task }} is succeeded`. This removes the boilerplate of having to register the output and check it succeeded. It also removes the need for the warning in the docs that simply specifying `retries` will not retry the task on failure.

Original
```yaml
  - name: old way
    get_url:
      url: http://1.2.3.4/zzz
      dest: /tmp/zzz
    retries: 2
    delay: 3
    register: result
    until: result is succeeded
# ---
  - name: new way
    get_url:
      url: http://1.2.3.4/zzz
      dest: /tmp/zzz
    retries: 2
    delay: 3
```
fixes #20802

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
until
TaskExecutor

##### ADDITIONAL INFORMATION
The existing MR #43128 appears to be inactive. One of the main reasons for returning it for rework was a lack of tests, for both the new and (especially) the existing behaviour. This MR  addresses that of those by extracting functions and pinning down the existing behaviour. It then changes the tests necessary for the new behaviour and demonstrates that they are compatible except for the new functionality. That is, it no longer ignores the `retries` parameter if `until` is not specified.

lilatomic committed:
extract method for processing retry arguments

lilatomic committed:
pull 'until' into process_retry_parameters

lilatomic committed:
honor `retries` even if `until` is absent

this is one step to allowing a bare `retries` to default
to using `until: {{ task }} is succeeded`.
We need to be able to distinguish whether `retries` has been manually specified.
This requires 2 changes:
  - removing the default for the variable
  - assuming that `retries: null` is equivalent to not specifying a value for `retries` (or vice versa)

So I have removed the default for that variable.
Removing the default value is acceptable to me.
It's use here is overridden and it isn't used elsewhere.

One consequence of treating `retries: null` as unspecified
is that will now result in 4 attempts, and not 3.
(It is now not possible to distinguish those cases)
I think that's fine, as `retries: null` is either nonsensical
or an explicit request for default.

The alternative is to use a sentry value for the default,
but that seems the _most_ risky option,
since the type will not match the expected type
but will not evaluate as None

lilatomic committed:
rename `retries`->`attempts`

I think this explains why the `+1` pops up

lilatomic committed:
pull `until` out of process_retry_parameters

Creating the `Conditional` object and evaluating the current status
would require too much knowledge from outside the task.
I don't think that's the way to go.

lilatomic committed:
extract method to determine retries requested

There are 3 different criteria used to test if this task wants retries,
some of which rely on implicit behaviour
of the function which calculates retry numbers.
This pulls it all together and simplifies downstream logic.

lilatomic committed:
simplify process_retry_parameters

usign flag for retries_requested

lilatomic committed:
use retries_requested flag

lilatomic committed:
add test for `retries` without `until`

lilatomic committed:
implement defaulting `until` to task succeeded

lilatomic committed:
cleanup tests

lilatomic committed:
document ability to specify `retries` without `until`

lilatomic committed:
add changelog fragment

lilatomic committed:
lint

s-hertel commented on lib/ansible/playbook/task.py:
```diff
@@ -82,7 +82,7 @@ class Task(Base, Conditional, Taggable, CollectionSearch):
     _poll = FieldAttribute(isa='int', default=C.DEFAULT_POLL_INTERVAL)
     _register = FieldAttribute(isa='string', static=True)
-    _retries = FieldAttribute(isa='int', default=3)
+    _retries = FieldAttribute(isa='int')
```
It looks like this change is already tested. We already have an integration test that makes sure `until` without `retries` makes 3 attempts. :+1:

s-hertel commented on test/units/executor/test_task_executor.py:
```diff
@@ -17,9 +17,13 @@
 from __future__ import (absolute_import, division, print_function)
+from typing import Dict
+
+from lib.ansible.playbook.task import Task
```
this should be  `from ansible.playbook.task`

lilatomic replied:
serves me right for trusting autoimport

s-hertel commented on lib/ansible/executor/task_executor.py:
```diff
@@ -770,6 +763,31 @@ def _evaluate_failed_when_result(result):
+                attempts += 3
+
+        if attempts <= 0:  # guard against negative retries
+            attempts = 1
```
It doesn't look like we have coverage for this codepath.

lilatomic replied:
true, done. While testing for this, I also added a testcase for `retries: 0`, and found that I was caught up by `0` being falsy (so it would fall through to the "retries unspecified" case and result in 4, not 1)
Sweet Catch

s-hertel commented on lib/ansible/executor/task_executor.py:
```diff
@@ -770,6 +763,31 @@ def _evaluate_failed_when_result(result):
+            attempts = 1
+
+        delay = _task.delay
+        if delay < 0:
```
We don't have coverage for this either. Since testing this would be for backwards compatiblity, please add integration tests (not unit tests - the unit tests are good but they all rely on the new TE method).

lilatomic replied:
true, done. I disagree with the necessity for an integration test. The unit tests don't actually rely on the _new_ TE method. They are added in the first commit, which only extracts code. So they are tests of the current behaviour. Well, with the possibility for some scoping quirks, but we can check for those. Now, proving that I've used refactoring techniques and that I've looked for those scoping things is an entirely other matter (especially with the 'new_contributor' label haha). So I've just added the integration tests.

Is there a good way for me to signal that some commits are refactoring groundwork?

s-hertel replied:
The reason I asked for integration tests is because it's easy to check backwards compatibility. I can't copy your unit tests onto devel without also editing code since they use the new method.

s-hertel commented on lib/ansible/executor/task_executor.py:
```diff
@@ -558,21 +559,8 @@ def _execute(self, variables=None):
-        # Read some values from the task, so that we can modify them if need be
-        if self._task.until:
-            retries = self._task.retries
-            if retries is None:
```
Removing this is a small change in behavior, but seems like an improvement. Before if `until` was provided with `retries: ~ `, retries was set to 3 (instead the default += 1), resulting in one fewer attempts. This could use test coverage to keep the behavior consistent going forward.

lilatomic replied:
oh, wild. I didn't think there would be a difference  between deserialising a task from 
- `{"until": self.valid_until}`, which results in 4
- `{"until": self.valid_until, "retries": None}`, which results in 3

good catch!

If I read your message correctly, we'd like it to always default to 4, and that I should  add a test that it does that.
I hope it's that, because I don't see a clean way to distinguish between a manually specified `null` and an unspecified default to `None`.  I blather on about this in [this commit message](https://github.com/ansible/ansible/pull/76101/commits/95790252fcfc42e336c7eec36c150c44573b90fe)

s-hertel replied:
That is what I meant, It's a minor change but before it was a weird inconsistency. :-) Thanks for adding a test for it.

s-hertel commented on lib/ansible/executor/task_executor.py:
```diff
@@ -770,6 +763,31 @@ def _evaluate_failed_when_result(result):
+        if attempts <= 0:  # guard against negative retries
+            attempts = 1
+
+        delay = _task.delay
```
We're missing coverage that the delay defaults to 5.

lilatomic replied:
`test_process_retry_parameters_no_delay_defaults_to_Task_default` tests that, although it uses the `Task._delay.default` parameter. Should I explicitly pin it to that value?

s-hertel replied:
Oh, that looks great. Nevermind me :smile:

s-hertel reviewed (commented):
Thanks for working on this. I left some comments about coverage that seems to be missing for the behavior on devel to ensure backwards compatibility. I also don't see any tests using variables for delay/retries, though I might have missed it in another test suite.

lilatomic committed:
`delay: null` uses default

instead of crashing

lilatomic committed:
Add more integration tests

- null value for `delay`
- variables for retry parameters

samccann commented on changelogs/fragments/76101-retries-without-until.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - using ``retries`` without specifying ``until`` defaults to retrying until the task is successful
```
Please add a link to the issue here. It helps readers go back and dig into the details if they want  to.

samccann commented on docs/docsite/rst/user_guide/playbooks_loops.rst:
```diff
@@ -268,7 +268,15 @@ To see the results of individual retries, run the play with ``-vv``.
 When you run a task with ``until`` and register the result as a variable, the registered variable will include a key called "attempts", which records the number of the retries for the task.
 
-.. note:: You must set the ``until`` parameter if you want a task to retry. If ``until`` is not defined, the value for the ``retries`` parameter is forced to 1.
+.. versionadded:: 5.0
```
```suggestion
.. versionadded:: 2.13
```

samccann commented on docs/docsite/rst/user_guide/playbooks_loops.rst:
```diff
@@ -268,7 +268,15 @@ To see the results of individual retries, run the play with ``-vv``.
-.. note:: You must set the ``until`` parameter if you want a task to retry. If ``until`` is not defined, the value for the ``retries`` parameter is forced to 1.
+.. versionadded:: 5.0
+
+In the common case of retrying a task until it is successful, you can use ``retries`` without ``until``. For example, 
```
```suggestion
In the common case of retrying a task until it is successful, you can use ``retries`` without ``until``. For example:
```

samccann reviewed (commented):
a few docs suggestions...

sivel commented on lib/ansible/executor/task_executor.py:
```diff
@@ -13,6 +13,7 @@
 import sys
 import termios
 import traceback
+from typing import List, Tuple, Union
```
We haven't defined how we want to do type annotations as of yet in the controller code base.  As such, I think we should remove it from this PR.

lilatomic replied:
I've removed the typehinting from the core and shuffled it into the same format I saw elsewhere. 

I don't see an open issue in ansible-community/community-topics about type hints, should I create one?

bcoca replied:
since there are many competing standards we decided to stop all type hinting being added until we choose one, then only that one will be allowed for core code.

lilatomic replied:
yep. this thread is why I created [the item in ansible/community](https://github.com/ansible/community/issues/635#issuecomment-966636254)

lilatomic committed:
suggested doc improvements

- use version of correct artefact (ansible-core, not ansible)
- add link to issue in change fragment
- fix punctuation in doc

lilatomic committed:
turn type hints into comments

awaiting formalisation of typehints

samccann reviewed (approved):
Docs portion LGTM!

s-hertel reviewed (approved):
The test coverage looks great now, thanks.

lilatomic committed:
fixup typehint-like comments

webknjaz commented:
/azp run

bcoca commented:
@lilatomic can you address the remaining concerns
waiting_on_contributor

bcoca commented on lib/ansible/executor/task_executor.py:
```diff
@@ -824,6 +817,33 @@ def _evaluate_failed_when_result(result):
+
+        delay = _task.delay
+        if delay is None:
+            delay = _task._delay.default
```
this wont work for 2 reasons a) you removed the default from the definition b) field attributes have changed so this is accessing a non existant method

lilatomic replied:
I'm not sure I understand what you're saying. The default for the `default` parameter is `None`. The attribute I removed the default from is `_retries`, not `_delay`. There is already a test that a task with an `until` section but not `retries` has the default of 3, and it is still passing.
But yeah, I guess a lot changes in a year. I'll see if this rebases.

lilatomic replied:
oh, the new implementation of `FieldAttribute` doesn't support introspection. The builtin `Property` has this introspectability:
```python
>>> class C:
...     @property
...     def p(self):
...         ...
... 
>>> C.p
<property object at 0x7fc10158b6f0>
```
adding this is simple, in `FieldAttribute.__get__` we just need the line [ref](https://docs.python.org/3/howto/descriptor.html#properties)
```python
    if obj is None:
        return self
```
This enables programatically accessing the default of the `FieldAttribute`, making tests less brittle. Alternatively, `FieldAttribute` coincidentally returns the default if invoked when unbound, so I can just access that in a test and hope it doesn't break in the future. Or I can just hardcode the expected value of the default and things will break if that ever changes.

bcoca commented on changelogs/fragments/76101-retries-without-until.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
```
not sure this is 'minor' .. also an entry in the porting guide will probably be needed

lilatomic replied:
It's minor by the semver definition because it adds functionality in a backwards-compatible way. None of the existing tests needed to be modified. If you have ideas for situations where the behaviour might have been modified, I can add a test and find out.

lilatomic committed:
propagate param rename in test

lilatomic committed:
access FieldAttribute, not attribute

lilatomic committed:
FieldAttribute supports introspection

bcoca commented on lib/ansible/playbook/attribute.py:
```diff
@@ -155,6 +155,8 @@ def __init__(self, extend=False, prepend=False, **kwargs):
 
     def __get__(self, obj, obj_type=None):
+        if obj is None:
+            return self
```
@mkrizek is this OK with your planned changes?

mkrizek replied:
I suppose it does not break anything but trying to understand why is it needed for this PR in the comment above.

bcoca replied:
i think the main issue is detecting 'unset' for this keyword, which having a default makes harder

mkrizek commented on lib/ansible/executor/task_executor.py:
```diff
@@ -817,6 +810,33 @@ def _evaluate_failed_when_result(result):
+
+        delay = _task.delay
+        if delay is None:
+            delay = Task.delay.default
```
This does not seem right to me. The default should be automatically set by the field attribute, no?

lilatomic replied:
added in [this commit](https://github.com/lilatomic/ansible/commit/0e408e15159cfce2ed88543e72a4e79202acaf1b) to support `delay: null`, in response to [this comment](https://github.com/ansible/ansible/pull/76101#discussion_r733923981) after discovering that not all invalid delays would be corrected to valid delays.

mkrizek commented:
!waffling needs_revision

webknjaz commented on test/units/executor/test_task_executor.py:
```diff
@@ -17,14 +17,18 @@
 
 # Make coding more python3-ish
 from __future__ import (absolute_import, division, print_function)
+from typing import Dict
```
Maybe put this under "if type checking" since it's not even used as a literal?

lilatomic replied:
I don't see `if TYPE_CHECKING` used anywhere else in the ansible codebase. With regards to typechecking, I moved them into comments at the request [here](https://github.com/ansible/ansible/pull/76101#discussion_r744211380), and created [this issue](https://github.com/ansible/community/issues/635#issuecomment-966636254) in ansible/community. From [this comment](https://github.com/ansible/proposals/issues/44#issuecomment-1273388961) on a proposal, it looks like we're using the builtin typehint format now. Should I add the comments back as typehints?

s-hertel commented:
Resolved by #81531
